File: SudokuParser.java
import java.io.*;
import java.util.*;

public class SudokuParser {

  private static final String LINE_SEPARATOR = "+=======================+";
  private static final int REGION_SIZE = 3;

  // takes a sudoku puzzle and writes it to a file
  public static void writeSudoku(Square[] puzzle, String fileName) throws IOException {
    try(BufferedWriter bw = new BufferedWriter(new FileWriter(fileName))) {
      for(int square_no = 0; square_no < puzzle.length; square_no++ ){

        // write a new line if we've written a full row
        if(square_no%Square.NUM_SQUARES_ROW_COL==0 && square_no != 0) {
          bw.write("| ");
          bw.newLine();
        }

        // write a horizontal bar to separate regions
        if(square_no%(Square.NUM_SQUARES_ROW_COL * REGION_SIZE) == 0) {
          bw.write(LINE_SEPARATOR);
          bw.newLine();
        }

        // write a pipe to separate the regions
        if(square_no%REGION_SIZE == 0) {
          bw.write("| ");
        }

        String numberToWrite = Integer.toString(puzzle[square_no].get_value());
        if(puzzle[square_no].get_value() == Square.UNKNOWN) {
          bw.write("- ");
        } else {
          bw.write(numberToWrite + " ");
        }
      }

      bw.write("| ");
      bw.newLine();
      bw.write(LINE_SEPARATOR);
      bw.close();
    }
  }

  // Reads in a sudoku grid from a text file and converts it to a square array
  public static Square[] parseSudoku(String fileName) throws IOException {

    List<Square> squareList = new ArrayList<Square>();

    // open the file and read line by line
    // skip lines that start with the '+' character
    // remove the pipes and convert the '-' to '0'
    // split the string by spaces then get each int value for the square
    try(BufferedReader br = new BufferedReader(new FileReader(fileName))) {
      int square_no = 0;
      for(String line; (line = br.readLine()) != null; ) {
          // if it's a non-spacing line, then convert to squares
          if(line.length() > 0 && (line.charAt(0) != '+')) {
            line = line.replace("|", "").replace("-", "0");
            String[] values = line.split(" ");

            for(int i=0; i < values.length; i++) {
              if(values[i].length() > 0) {
                try {
                  squareList.add(new Square(square_no, Integer.parseInt(values[i])));
                } catch (NumberFormatException nfe) {
                  System.out.println("Error in parsing Sudoku from file" + nfe);
                  System.out.println("Unable to convert grid values.");
                } finally {
                  square_no++;
                }
              }
            }
          }
      }

      // after parsing the file, return the squareList
      return squareList.toArray(new Square[squareList.size()]);
    }
  }

  private static void printStringArray(String[] strings) {
    for(String string: strings) {
      System.out.print(string + " " );
    }
    System.out.print("\n");
  }

}

File: cpp/Sudoku/Sudoku/Sudoku_Parser.cpp
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <math.h>
#include <string>
#include <sstream>
#include <stdio.h>
#include <set>
#include <algorithm>

#include "Square.h"
#include "Sudoku_Parser.h"

using namespace std;

const string Sudoku_Parser::LINE_SEPARATOR = "+=======================+";

//Sudoku_Parser
//write_sudoku method takes an array of Square objects and prints the value of each Square in proper
//sudoku grid format to the file specified
//parseSudoku takes the file given and attempts to parse the sudoku puzzle in the file into
//an array of Square objects 

//@authors Matt Hino and Christine Chen
//@date 10/1/2016

int Sudoku_Parser::write_sudoku(Square* puzzle, string fileName) {

	ofstream myfile(fileName);

	// if we were unable to open the file, exit out gracefully
	if (!myfile.is_open()) {
		return Sudoku_Parser::ERROR;
	}

	// opened file nicely (:
	for (int square_no = 0; square_no < Square::TOTAL_NUM_SQ; square_no++){

		// write a new line if we've written a full row
		if (square_no%Square::NUM_SQ_ROW_COL == 0 && square_no != 0) {
			myfile << "| " << endl;
		}

		// write a horizontal bar to separate regions
		if (square_no % (Square::NUM_SQ_ROW_COL * Sudoku_Parser::REGION_SIZE) == 0) {
			myfile << Sudoku_Parser::LINE_SEPARATOR << endl;
		}

		// write a pipe to separate the regions
		if (square_no%Sudoku_Parser::REGION_SIZE == 0) {
			myfile << "| ";
		}

		int numberToWrite = (puzzle + square_no) -> get_value();
		if (numberToWrite == Square::UNKNOWN) {
			myfile << "- ";
		}
		else {
			myfile << numberToWrite << " ";
		}
	}

	myfile << "| " << endl << Sudoku_Parser::LINE_SEPARATOR;
	myfile.close();
	return 0;
}

/*
* This method reads in a sudoku file, creates, and intializes an array of squares
* then returns a pointer to that square array.
*/
Square* Sudoku_Parser::parseSudoku(string file_name) {

	ifstream myfile(file_name);
	// make sure to exit if we aren't able to open the file nicely.
	if (!myfile.is_open()) {
		return nullptr;
	}

	Square* square_arr = new Square[Square::TOTAL_NUM_SQ];
	string line;
	int square_no = 0;

	while (getline(myfile, line)) {
		if (line.size() > 0 && (line[0] != '+')) {

			// get rid of spacing characters and dashes
			replace(line.begin(), line.end(), '|', ' ');
			replace(line.begin(), line.end(), '-', '0');

			// iterate through the string, if we come across a number, then add a square
			// with that value to the square array.
			for (int i = 0; i < line.size(); i++) {
				if (line[i] != ' ') {
					int value = atoi(&line[i]);
					*(square_arr + square_no) = Square(square_no, value);
					square_no++;
				}
			}
		}
	}

	myfile.close();
	return square_arr;
}


